File: Numero_Complexo.rb
class Numero_Complexo

  def initialize (real, imaginaria)
    @real = real
    @imaginaria = imaginaria
  end

  attr_reader :real, :imaginaria

  def somar(outro_numero)
    Numero_Complexo.new(@real + outro_numero.real, @imaginaria + outro_numero.imaginaria)
  end

  def subtrair(outro_numero)
    Numero_Complexo.new(@real - outro_numero.real, @imaginaria - outro_numero.imaginaria)
  end

  def multiplicar(outro_numero)
    Numero_Complexo.new((@real*outro_numero.real) - (@imaginaria*outro_numero.imaginaria), 
                        (@imaginaria*outro_numero.real) + (@real*outro_numero.imaginaria))
  end

  def retorna_string()
    "#{@real} + #{@imaginaria}i"
  end

  def ==(outro_numero)
    @real == outro_numero.real && @imaginaria == outro_numero.imaginaria
  end

  def parte_real()
    @real
  end

  def parte_imaginaria()
    @imaginaria
  end
end

File: NumeroComplexo.java

public class NumeroComplexo {

	private double real;
	private double imaginaria;

	public NumeroComplexo(double real, double imaginaria) {
		this.real = real;
		this.imaginaria = imaginaria;
	}

	public double real() {
		return this.real;
	}
	
	public double imaginaria() {
		return this.imaginaria;
	}

	public NumeroComplexo somar(NumeroComplexo outroNumero) {
		return new NumeroComplexo(this.real + outroNumero.real(), 
								  this.imaginaria + outroNumero.imaginaria());
	}
	
	public boolean equals(Object outro){
		NumeroComplexo outroNumero = (NumeroComplexo) outro;
		return this.real == outroNumero.real() && this.imaginaria == outroNumero.imaginaria; 
	}

	public NumeroComplexo subtrair(NumeroComplexo outroNumero) {
		return new NumeroComplexo(this.real - outroNumero.real(), 
				  this.imaginaria - outroNumero.imaginaria());
	}

	public NumeroComplexo multiplicar(NumeroComplexo outroNumero) {
		return new NumeroComplexo((this.real * outroNumero.real()) - (this.imaginaria * outroNumero.imaginaria()), 
				  (this.imaginaria * outroNumero.real()) + (this.real * outroNumero.imaginaria()));
	}

	public String retornarString() {
		return String.format("%.0f + %.0fi", this.real, this.imaginaria);
	}
}
